### PR TITLE
Added elastic transform in torchvision.transforms.

### PR DESCRIPTION
**Pull request by gokriznastic**
I've added elastic transformations support as requested in #888.

### CODE REVIEW

File: torchvision/transforms/functional.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 import sys
 import math
 from PIL import Image, ImageOps, ImageEnhance, PILLOW_VERSION
+from scipy.ndimage.interpolation import map_coordinates
+from scipy.ndimage.filters import gaussian_filter
```
Comment (pmeier) [inline torchvision/transforms/functional.py]: This makes `torchvision` dependent on `scipy` at import. Up until now we avoided this, so we should discuss if we want to have this now.

File: torchvision/transforms/functional.py [hunk 2]
```diff
@@ -855,3 +857,45 @@ def erase(img, i, j, h, w, v, inplace=False):
...
+        sigma (float): Elasticity coefficient as described in [Simard2003]
+        random_state (int): Random state to initialize the Gaussian kernel
+
+    Returns:
+        PIL Image: Image with elastic distortions.
+    """
+    if not _is_pil_image(img):
+        raise TypeError('img should be PIL Image. Got {}'.format(type(img)))
+
+    img = np.asarray(img)
```
Comment (pmeier) [inline torchvision/transforms/functional.py]: This conversion dismisses that the PIL image has different modes and has to be scaled and converted accordingly. You could use `img = to_tensor(img).numpy()` if you don't want to handle the correct conversion yourself.

File: torchvision/transforms/functional.py [hunk 3]
```diff
@@ -855,3 +857,45 @@ def erase(img, i, j, h, w, v, inplace=False):
...
+        raise ValueError('pic should be 2/3 dimensional. Got {} dimensions.'.format(img.ndim))
+
+    if len(img.shape) < 3:
+        img = img.reshape(img.shape[0], img.shape[1], -1)
+
+    shape = img.shape
+
+    dx = gaussian_filter((random_state.rand(*shape) * 2 - 1), sigma, mode="constant", cval=0) * alpha
+    dy = gaussian_filter((random_state.rand(*shape) * 2 - 1), sigma, mode="constant", cval=0) * alpha
+    dz = np.zeros_like(dx)
```
Comment (pmeier) [inline torchvision/transforms/functional.py]: Using `numpy` or `scipy` as base for the calculations also prevents tensors on the GPU to be processed

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -325,6 +325,43 @@ def __repr__(self):
...
+    Args:
+        alpha (float): Scaling factor as described in [Simard2003] Default value is 100
+        sigma (float): Elasticity coefficient as described in [Simard2003] Default value is 10
+        random_state (int, optional): Random state to initialize the Gaussian kernel
+
+    """
+
+    def __init__(self, alpha=100, sigma=10, random_state=None):
+        if random_state is None:
+            self.random_state = np.random.RandomState(None)
```
Comment (pmeier) [inline torchvision/transforms/transforms.py]: Is the `RandomState` necessary? Couldn't you take a `seed` and create the random values without a specific `RandomState`?

File: torchvision/transforms/functional.py (hunk 1 above)
Comment (gokriznastic) [reply torchvision/transforms/functional.py]: Firstly, thank you for the quick feedback. 
To answer your question, in all the implementations of elastic transform that I've seen `scipy.ndimage.filters.gaussian_filter` and `scipy.ndimage.interpolation.map_coordinates` were being used unanimously. I understand that this makes `torchvision` dependent on `scipy`, but I have not found any workaround to this yet.

File: torchvision/transforms/functional.py [hunk 5]
```diff
@@ -855,3 +857,45 @@ def erase(img, i, j, h, w, v, inplace=False):
...
+
+def elastic_transform(img, alpha, sigma, random_state):
+    """Elastic deformation of image as described in [Simard2003]_.
+    .. [Simard2003] Simard, Steinkraus and Platt, "Best Practices for
+       Convolutional Neural Networks applied to Visual Document Analysis", in
+       Proc. of the International Conference on Document Analysis and
+       Recognition, 2003.
+
+    Args:
+        img (PIL Image): Image on which elastic transformation is to be applied
```
[the remark this answers is not in the document]
Comment (gokriznastic) [reply torchvision/transforms/functional.py]: This was done just to maintain consistency with other `transforms` available, as most `transforms` in `torchvision` work on a `PIL.Image`. or `torch*.Tensor`. Which is the better way to go about this in your opinion?

File: torchvision/transforms/functional.py (hunk 2 above)
Comment (gokriznastic) [reply torchvision/transforms/functional.py]: Yes, thank you for pointing this out. That'd definitely be the better way to convert the image into `numpy`.

File: torchvision/transforms/functional.py (hunk 1 above)
Comment (pmeier) [reply torchvision/transforms/functional.py]: You could use create a gaussian filter yourself with the [corresponding equations](https://en.wikipedia.org/wiki/Multivariate_normal_distribution) and replace `map_coordinates` with a variant of [`grid_sample`](https://pytorch.org/docs/stable/nn.functional.html?highlight=grid%20sample#torch.nn.functional.grid_sample). This way no `numpy` or `scipy`.

Comment (gokriznastic) [reply torchvision/transforms/functional.py]: @pmeier Thanks for pointing these out. I'll look into these and get back to you.

Comment (suvojit-0x55aa) [reply torchvision/transforms/functional.py]: `gaussian_filter` can be implemented with `Conv` and gaussian weights, here's [the relevant discussion](https://discuss.pytorch.org/t/is-there-anyway-to-do-gaussian-filtering-for-an-image-2d-3d-in-pytorch/12351/7)  

As for map_coordinates, is similar to [`grid_sample`](https://pytorch.org/docs/stable/nn.functional.html?highlight=grid%20sample#torch.nn.functional.grid_sample) as @pmeier said. The interpolation available for `grid_sample` is `bilinear` and `nearest` . Where as `map_coordinates` is interpolated using spline interpolation. If my understanding is correct implementing spline interpolation in `grip_sample` should do the job.

**Comment by gokriznastic** [comment]
@fmassa thank you for the suggestions. I'm working on it. Might take some time.

**Comment by fmassa** [comment]
Sure, no worries, thanks for the PR!

**Comment by Randl** [comment]
@gokriznastic you need some help with it?
